llm/memory.go

```diff
@@ -0,0 +1,120 @@
+package llm
+
+import (
+	"fmt"
+	"log/slog"
+
+	"github.com/ollama/ollama/api"
+	"github.com/ollama/ollama/format"
+	"github.com/ollama/ollama/gpu"
+)
+
+// This algorithm looks for a complete fit to determine if we need to unload other models
+func PredictServerFit(allGpus gpu.GpuInfoList, ggml *GGML, adapters, projectors []string, opts api.Options) (bool, uint64, error) {
+	var estimatedVRAM uint64
+	var layers int
+	if opts.NumCtx > int(ggml.KV().ContextLength()) {
+		slog.Warn("requested context length is greater than model max context length", "requested", opts.NumCtx, "model", ggml.KV().ContextLength())
+		opts.NumCtx = int(ggml.KV().ContextLength())
+	}
+
+	if opts.NumCtx < 4 {
+		opts.NumCtx = 4
+	}
+
+	// Split up the GPUs by type and try them
+	for _, gpus := range allGpus.ByLibrary() {
+		layers, estimatedVRAM = PredictGPULayers(gpus, ggml, projectors, opts)
+		if layers >= int(ggml.KV().BlockCount()+1) || (opts.NumGPU > 0 && layers >= opts.NumGPU) {
```

Comment:
This seems like a brittle check since opts.NumGPU is changed by PredictGPULayers. Perhaps PredictGPULayers should return true/false if the model fits fully on the GPU?

Reply:
Good catch.

Reply:
I realize now `PredictGPULayers` passes opts by value so this might be fine